### PR TITLE
logger callbacks

### PR DESCRIPTION
**Pull request by geyslan**
commit 761bb030fcb998dfe9915601a538f6087fb6d978

    selftest: add log-callbacks

commit 56803f4389f7557a47eba7e7e78015cddefcf424

    logger: introduce logger callbacks
    
    loggerCallback() calls callbacks, log() and []logFnFilters(),
    which can be set by the libbpfgo consumer via SetLoggerCbs().
    
    This moves output filtering from C libbpf_print_fn() to
    Go filterOutput() which can be set as one of the slice of filter funcs.
    
    This also introduces LogWarnLevel, LogInfoLevel and LogDebugLevel
    constants.

Fixes: #276
Fixes: #6 
Fixes: #55 
Fixex: https://github.com/aquasecurity/tracee/issues/2417

- https://github.com/aquasecurity/tracee/issues/2417
- #276
- #6
- #55

Related fix during the work:

- #285 

# Tests

```shell
libbpfgo/selftest/log-callbacks on  2417-logger-callback [$!?]
❯ make
make -C /home/gg/code/libbpfgo libbpfgo-static
make[1]: Entering directory '/home/gg/code/libbpfgo'
CC=clang \
        CGO_CFLAGS="-I/home/gg/code/libbpfgo/output" \
        CGO_LDFLAGS="-lelf -lz /home/gg/code/libbpfgo/output/libbpf.a" \
        GOOS=linux GOARCH=amd64 \
        go build \
        -tags netgo -ldflags '-w -extldflags "-static"' \
        .
...

libbpfgo/selftest/log-callbacks on  2417-logger-callback [$!?]
❯ sudo ./main-static                     

libbpfgo/selftest/log-callbacks on  2417-logger-callback [$!?]
❯ echo $?                                
0
```

Changing one filter callback for tests purpose only (== instead of !=):

```go
func(libLevel int, msg string) bool {
    return libLevel == C.LIBBPF_WARN
},
```

```shell
tracee on  main [$!?⇡] via 🐹 v1.19.5 via ⍱ v2.3.4 took 3s 
❯ sudo ./dist/tracee -t uid=1000 -t comm=who --log warn 2>&1 | head -20
{"level":"warn","ts":1675175312.2060382,"msg":"libbpf: loading object 'embedded-core' from buffer"}
{"level":"warn","ts":1675175312.2061195,"msg":"libbpf: elf: section(3) raw_tracepoint/sys_enter, size 248, link 0, flags 6, type=1"}
{"level":"warn","ts":1675175312.2061718,"msg":"libbpf: sec 'raw_tracepoint/sys_enter': found program 'tracepoint__raw_syscalls__sys_enter' at insn offset 0 (0 bytes), code size 31 insns (248 bytes)"}
{"level":"warn","ts":1675175312.2062,"msg":"libbpf: elf: section(4) .relraw_tracepoint/sys_enter, size 32, link 223, flags 40, type=9"}
{"level":"warn","ts":1675175312.2062263,"msg":"libbpf: elf: section(5) raw_tracepoint/sys_enter_init, size 2368, link 0, flags 6, type=1"}
{"level":"warn","ts":1675175312.2062707,"msg":"libbpf: sec 'raw_tracepoint/sys_enter_init': found program 'sys_enter_init' at insn offset 0 (0 bytes), code size 296 insns (2368 bytes)"}
{"level":"warn","ts":1675175312.206299,"msg":"libbpf: elf: section(6) .relraw_tracepoint/sys_enter_init, size 160, link 223, flags 40, type=9"}
{"level":"warn","ts":1675175312.206323,"msg":"libbpf: elf: section(7) raw_tracepoint/sys_enter_submit, size 24904, link 0, flags 6, type=1"}
{"level":"warn","ts":1675175312.2063725,"msg":"libbpf: sec 'raw_tracepoint/sys_enter_submit': found program 'sys_enter_submit' at insn offset 0 (0 bytes), code size 3113 insns (24904 bytes)"}
{"level":"warn","ts":1675175312.2064219,"msg":"libbpf: elf: section(8) .relraw_tracepoint/sys_enter_submit, size 416, link 223, flags 40, type=9"}
{"level":"warn","ts":1675175312.2064548,"msg":"libbpf: elf: section(9) raw_tracepoint/sys_exit, size 312, link 0, flags 6, type=1"}
{"level":"warn","ts":1675175312.2065027,"msg":"libbpf: sec 'raw_tracepoint/sys_exit': found program 'tracepoint__raw_syscalls__sys_exit' at insn offset 0 (0 bytes), code size 39 insns (312 bytes)"}
{"level":"warn","ts":1675175312.2065287,"msg":"libbpf: elf: section(10) .relraw_tracepoint/sys_exit, size 32, link 223, flags 40, type=9"}
{"level":"warn","ts":1675175312.2065518,"msg":"libbpf: elf: section(11) raw_tracepoint/sys_exit_init, size 1200, link 0, flags 6, type=1"}
{"level":"warn","ts":1675175312.2065969,"msg":"libbpf: sec 'raw_tracepoint/sys_exit_init': found program 'sys_exit_init' at insn offset 0 (0 bytes), code size 150 insns (1200 bytes)"}
{"level":"warn","ts":1675175312.2066414,"msg":"libbpf: elf: section(12) .relraw_tracepoint/sys_exit_init, size 144, link 223, flags 40, type=9"}
{"level":"warn","ts":1675175312.2066822,"msg":"libbpf: elf: section(13) raw_tracepoint/sys_exit_submit, size 18976, link 0, flags 6, type=1"}
{"level":"warn","ts":1675175312.2067318,"msg":"libbpf: sec 'raw_tracepoint/sys_exit_submit': found program 'sys_exit_submit' at insn offset 0 (0 bytes), code size 2372 insns (18976 bytes)"}
{"level":"warn","ts":1675175312.2067814,"msg":"libbpf: elf: section(14) .relraw_tracepoint/sys_exit_submit, size 416, link 223, flags 40, type=9"}
{"level":"warn","ts":1675175312.206826,"msg":"libbpf: elf: section(15) raw_tracepoint/trace_sys_enter, size 8120, link 0, flags 6, type=1"}
```

### Context

- https://github.com/aquasecurity/tracee/pull/2600

### CODE REVIEW

File: logger_cb.go
Comment (mozillazg) [inline logger_cb.go:13]: How about define an `interface` instead of depend on `tracee`?

Comment (AlonZivony) [reply logger_cb.go:13]: I think it will be hard to use an interface here because we do need the logger levels.
We can use the `zap` levels, but it will make an ununified interface.
Maybe we should create a standalone package under Tracee (like `types`) for it, or event an new project.

Comment (geyslan) [reply logger_cb.go:13]: > Maybe we should create a standalone package under Tracee (like types) for it, or create a new project.

This was my initial intent way back. If this is ok, we can merge it in next, start to test it and do issue logger, as an own project, to a farther milestone. WDYT?

File: logger_cb.go
Comment (rafaeldtinoco) [inline logger_cb.go:57]: nit: Exclusivity misspelled.

File: logger_cb.go
Comment (rafaeldtinoco) [inline logger_cb.go:19]: nit: IMO we should stick with camel notation for golang code only.

File: logger_cb.go
Comment (rafaeldtinoco) [inline logger_cb.go:55]: nit: mispell.

File: logger_cb.go
Comment (rafaeldtinoco) [inline logger_cb.go:56]: this long links for github don't help and polute the code, would u mind removing them ?

File: logger_cb.go
Comment (rafaeldtinoco) [reply logger_cb.go:12]: I agree.

**Review by rafaeldtinoco — APPROVED** [review]
LGTM after the small changes I requested (if you agree).

File: logger_cb.go
Comment (geyslan) [reply logger_cb.go:56]: > this long links for github don't help and polute the code, would u mind removing them ?

What about?

```suggestion
	// triggered by: libbpf/src/nlattr.c->libbpf_nla_dump_errormsg()

```

This would help in future research.

File: logger_cb.go
Comment (geyslan) [reply logger_cb.go:12]: - https://github.com/aquasecurity/tracee/issues/2599

File: logger_cb.go
Comment (rafaeldtinoco) [reply logger_cb.go:56]: yep better, thanks!

File: logger_cb.go
Comment (geyslan) [reply logger_cb.go:12]: As https://github.com/aquasecurity/tracee/pull/2600#pullrequestreview-1253549004 concluded, we'll tackle this creating a libbpfgo.Logger interface.

Comment (geyslan) [reply logger_cb.go:12]: > As [aquasecurity/tracee#2600 (review)](https://github.com/aquasecurity/tracee/pull/2600#pullrequestreview-1253549004) concluded, we'll tackle this creating a libbpfgo.Logger interface.

@rafaeldtinoco thanks for reviewing this.

Returing to this soon.

**Comment by geyslan** [comment]
> LGTM after the small changes I requested (if you agree).

@rafaeldtinoco thanks for reviewing this.

> As [aquasecurity/tracee#2600 (review)](https://github.com/aquasecurity/tracee/pull/2600#pullrequestreview-1253549004) concluded, we'll tackle this creating a libbpfgo.Logger interface.

Returing to this soon.

File: logger_cb.go
Comment (geyslan) [reply logger_cb.go]: Done. Putting this as ready for review.

**Comment by geyslan** [comment]
Despite having passed all the tests, I need to define the callbacks on them to make them silent.

**Comment by geyslan** [comment]
> Despite having passed all the tests, I need to define the callbacks on them to make them silent.

Solved. Ready for review.

**Review by mozillazg — APPROVED** [review]
LGTM

File: logger_cb.go
Comment (NDStrahilevitz) [inline logger_cb.go:58]: Both of these can be in a struct, then passing nil for both would look like
`libbpfgo.SetLoggerCbs(SettingStructName{})` and not `libbpfgo.SetLoggerCbs(nil, nil)` which I think is easier to understand (empty struct => default setting compared to passing NULL)

**Review by NDStrahilevitz — CHANGES_REQUESTED** [review]
Could you please give an example of how this is used with different settings and what they entail? Because i'm not sure how this looks work out in the end.
I had to look at the equivalent tracee PR to understand the first setting, and I have no idea how the filtering would work, maybe tests could help here.

File: logger_cb.go
Comment (geyslan) [reply logger_cb.go:58]: > I had to look at the equivalent tracee PR to understand the first setting, and I have no idea how the filtering would work, maybe tests could help here.

I agree. A test example is welcome. Will work on that.

> Both of these can be in a struct, then passing nil for both would look like
libbpfgo.SetLoggerCbs(SettingStructName{}) and not libbpfgo.SetLoggerCbs(nil, nil) which I think is easier to understand (empty struct => default setting compared to passing NULL)

It makes sense. 👍🏼

File: logger_cb.go
Comment (NDStrahilevitz) [inline logger_cb.go:115]: Do not use the `Match` functions directly from `regexp`,  as they recompile the regex per call. Please compile the regex outside the function scope once then reuse it.

File: logger_cb.go
Comment (NDStrahilevitz) [inline logger_cb.go:132]: ditto

File: logger_cb.go
Comment (NDStrahilevitz) [inline logger_cb.go:149]: ditto

File: logger_cb.go
Comment (NDStrahilevitz) [inline logger_cb.go:117]:
```suggestion
		fmt.Fprintf(os.Stderr, "error filtering libbpf print output: %v", err)
```

File: logger_cb_test.go
Comment (NDStrahilevitz) [inline logger_cb_test.go:5]: Great thanks for this!

File: logger_cb_test.go
Comment (NDStrahilevitz) [inline logger_cb_test.go:12]: I believe we shouldn't filter anything internally by default. Consumers should choose their own filters, and since I assume most of these filters were hard coded for tracee, they should now be removed and set as filters from tracee itself.

So as for this test, we can test the same patterns but remove their hardcoding before. The test will first set the wanted pattern then test that they are correctly filtered.

**Review by NDStrahilevitz — CHANGES_REQUESTED** [review]

File: logger_cb_test.go
Comment (geyslan) [reply logger_cb_test.go:12]: > I believe we shouldn't filter anything internally by default. Consumers should choose their own filters, and since I assume most of these filters were hard coded for tracee, they should now be removed and set as filters from tracee itself.

I agree.

> So as for this test, we can test the same patterns but remove their hardcoding before. The test will first set the wanted pattern then test that they are correctly filtered.

Ok. I'll tackle this soon. Thanks.

File: selftest/log-callbacks/main.go
Comment (rafaeldtinoco) [inline selftest/log-callbacks/main.go:40]: nit: s/ouput/output

File: selftest/log-callbacks/main.go
Comment (rafaeldtinoco) [inline selftest/log-callbacks/main.go:41]: nit: s/ouput/output

File: logger_cb.go
Comment (rafaeldtinoco) [inline logger_cb.go:52]: A comment saying you're overwriting callbacks if SetLoggerCbs is called would be nice here. I also agree that you should make default LogFilterOuptut generic just like LogFilterLevel (and let tracee implement the output filtering). Maybe open an issue, merge this PR and solve the issue soon.

**Comment by rafaeldtinoco** [comment]
@geyslan minor nits and then we merge. LGTM.

File: logger_cb.go
Comment (geyslan) [reply logger_cb.go:52]: Yeah, I agree. Tackling this soon.

Comment (geyslan) [reply logger_cb.go:52]: @rafaeldtinoco,done.

**Comment by geyslan** [comment]
After merging this and https://github.com/aquasecurity/tracee/pull/2663, I'll open an issue in both projects to tackle the hard-coded filters move.

**Review by rafaeldtinoco — APPROVED** [review]
LGTM

**Comment by rafaeldtinoco** [comment]
Feel free to undraft tracee's change and move on there. Thanks!

**Pull request merged**
